File: c_learn/AVL.c
#include<stdio.h>
#include<stdlib.h>

typedef struct AVLT {
	struct AVLT* right;
	struct AVLT* left;
	int data;
	int height;
}AVL;
int getHeight(AVL* t)
{
	if (t == NULL)
		return 0;
	else
		return t->height;
}
AVL* LeftRotation(AVL* t)
{
	AVL* temp = t->right;
	t->right = t->right->left;
	temp->left = t;
	t->height = getHeight(t->left) > getHeight(t->right)
		? getHeight(t->left) + 1 :
		getHeight(t->right) + 1;
	temp->height = getHeight(temp->left) > getHeight(temp->right)
		? getHeight(temp->left) + 1 :
		getHeight(temp->right) + 1;
	return temp;
}

AVL* RightRotation(AVL* t)
{
	AVL* temp = t->left;
	t->left = t->left->right;
	temp->right = t;
	t->height = getHeight(t->left) < getHeight(t->right)
		? getHeight(t->left) + 1 :
		getHeight(t->right) + 1;
	temp->height = getHeight(temp->left) < getHeight(temp->right)
		? getHeight(temp->left) + 1 :
		getHeight(temp->right) + 1;
	return temp;
}

AVL* keepBalance(AVL* t)
{
	if (getHeight(t) <= 2)
		return t;
	if (getHeight(t->left) >= getHeight(t->right) + 2
		&& getHeight(t->left->left) > getHeight(t->left->right))
		return RightRotation(t);//需要左旋的情况
	else if (getHeight(t->right) >= getHeight(t->left) + 2
		&& getHeight(t->right->right) > getHeight(t->right->left))
		return LeftRotation(t);//需要右旋的情况
	else if (getHeight(t->left) >= getHeight(t->right) + 2
		&& getHeight(t->left->left) < getHeight(t->left->right))
	{
		t->left = LeftRotation(t->left);
		return RightRotation(t);
	}
	else if (getHeight(t->right) >= getHeight(t->left) + 2
		&& getHeight(t->right->left) < getHeight(t->right->left))
	{
		t->right = RightRotation(t->right);
		return LeftRotation(t);
	}
	else
		return t;
}
AVL* AddAVLT(int num, AVL* t)
{
	if (t == NULL)
	{
        AVL *newNode = (AVL*)malloc(sizeof(AVL));
		newNode->data = num;
        newNode->height = 1;
        newNode->left = NULL;
        newNode->right = NULL;
		return newNode;
	}
	if (num > t->data)
		t->right = AddAVLT(num, t->right);
	else
		t->left = AddAVLT(num, t->left);
	t->height = getHeight(t->left) > getHeight(t->right) ?
		getHeight(t->left) + 1:
		getHeight(t->right) + 1;
	return keepBalance(t);
}

int main()
{
	AVL T = { NULL, NULL, 13, 1 };
	for (int i = 0; i < 100000; i++)
	{
		int temp = i * (30 - i) * (i - i * (17 - i));
        // printf("%d",i);
		AddAVLT(temp, &T);
	}
    printf("%d\n", T.height);
}

File: go_test/Datastructure/AVL.go
package datastructure

import (
	"go_test/Tool"
)

type AVLT struct{
	Left *AVLT
	Right *AVLT
	val int
	height int
}

func (t *AVLT)LeftRotation() *AVLT{
	temp := t.Right
	t.Right = t.Right.Left
	temp.Left = t
	t.height = Tool.Max(t.Left.GetHeight(), t.Right.GetHeight())+1
	temp.height = Tool.Max(temp.Left.GetHeight(), temp.Right.GetHeight())+1
	return temp
}
func (t *AVLT)RightRotation() *AVLT{
	temp := t.Left
	t.Left = t.Left.Right
	temp.Right = t
	t.height = Tool.Max(t.Left.GetHeight(), t.Right.GetHeight())+1
	temp.height = Tool.Max(temp.Left.GetHeight(), temp.Right.GetHeight())+1
	return temp
}
func (t *AVLT)KeepBalance() *AVLT{
	if t.GetHeight() <= 2 {
		return t
	}
	if t.Left.GetHeight() >= t.Right.GetHeight() + 2 && t.Left.Left.GetHeight() > t.Left.Right.GetHeight() {
		return t.RightRotation()//LL
	}else if t.Right.GetHeight() >= t.Left.GetHeight() + 2 && t.Right.Right.GetHeight() > t.Right.Left.GetHeight(){
		return t.LeftRotation()//RR
	}else if t.Left.GetHeight() >= t.Right.GetHeight() + 2 && t.Left.Left.GetHeight() < t.Left.Right.GetHeight(){
		t.Left = t.Left.LeftRotation()
		return t.RightRotation()
	}else if t.Right.GetHeight() >= t.Left.GetHeight() + 2 && t.Right.Right.GetHeight() < t.Right.Left.GetHeight(){
		t.Right = t.Right.RightRotation()
		return t.LeftRotation()
	}else{
		return t
	}
}
func (t *AVLT)Add(k int) *AVLT{
	if t == nil {
		newNode := AVLT{nil,nil, k, 1}
		return &newNode
	}
	if k > t.val {
		t.Right = t.Right.Add(k)
	}else{
		t.Left = t.Left.Add(k)
	}
	t.height = 1 + Tool.Max(t.Left.GetHeight(), t.Right.GetHeight())
	return t.KeepBalance()
}
func (t *AVLT)Print(){//输出节点
	if t != nil {	
		t.Left.Print()
		println(t.val)
		t.Right.Print()
	}
}

func (t *AVLT)GetHeight() int{
	if t == nil {
		return 0
	}
	return t.height
}
func (t *AVLT)SetHeight(k int){
	t.height = k
}
func (t *AVLT)GetVal() int{
	if t == nil {
		return 0
	}
	return t.val
}
func (t *AVLT)SetVal(k int){
	t.val = k
}
func (t *AVLT)Height() int{
	if(t == nil){
		return 0
	}
	return Tool.Max(t.Left.GetHeight()+1, t.Right.GetHeight()+1)
}
func NewAVLT(k int) *AVLT{
	t := AVLT{nil, nil, k, 1}
	return &t
}